### ISSUE TITLE
Copying does not preserve fog_attributes

### ISSUE THREAD

**Issue by vcsjones** [carrierwaveuploader/carrierwave#2240]
We are using CarrierWave with S3 where our `cache_storage` is also set to `:fog`  so that temporary files work in S3. I won't go in to the details why, but it has been working to solve a business requirement.

We also started using KMS encryption and set the KMS key properties accordingly:

```ruby
c_config.fog_attributes = {
  encryption: 'aws:kms',
  kms_key_id: 'blah blah key id'
}
```

This work correctly when `cache_storage` is *not* set to fog. When it is set to fog, we see in S3 that the final document does not have the KMS key property set. However, the file in the cache directory in S3 does have the KMS key set. So the KMS key is lost when it is copied from the cache location to the final destination.

After a bit of debugging I believe I have figured out why.

Over at 

https://github.com/carrierwaveuploader/carrierwave/blob/e9f3be59e6e6b5d41a9b379df92cad6be16e7f84/lib/carrierwave/storage/fog.rb#L310-L313

The `store` operation uses the fog provider's native "copy" functionality if the copy is from the same provider to another. So the inclusion of the `fog_attributes` is completely missed from the `else`:

https://github.com/carrierwaveuploader/carrierwave/blob/e9f3be59e6e6b5d41a9b379df92cad6be16e7f84/lib/carrierwave/storage/fog.rb#L320

[Amazon states](http://docs.aws.amazon.com/AmazonS3/latest/dev/CopyingObjectsExamples.html) that the CopyObject command itself must also specify the KMS key:

>When copying objects you can request Amazon S3 to save the target object encrypted using an AWS Key Management Service (KMS) encryption key, an Amazon S3-managed encryption key, or a customer-provided encryption key. 

The fix is here:

https://github.com/carrierwaveuploader/carrierwave/blob/e9f3be59e6e6b5d41a9b379df92cad6be16e7f84/lib/carrierwave/storage/fog.rb#L414

Where the last parameter is `acl_header`. If you merge that with the `fog_attributes`, then the KMS key is preserved in to the copy command.

**Comment by vcsjones** [comment]
I should say, that it isn't quite as simple as merging the properties with `acl_header`. When using `kms_key_id` in the fog configuration, it appropriately translates this to the right HTTP header `x-amz-server-side-encryption-aws-kms-key-id` when a file is uploaded. However, the `copy_object` command does not perform this translation. Thus only adding `x-amz-server-side-encryption-aws-kms-key-id` to the hash will have the desired affect.

**Comment by mshibuya** [comment]
Fixed by #2503.
